cmds/tail/tail.go

```diff
@@ -0,0 +1,251 @@
+		if pos == 0 {
+			break
+		}
+		var thisChunkSize int64
+		if pos < blkSize {
+			thisChunkSize = pos
+		} else {
+			thisChunkSize = blkSize
+		}
+		pos -= thisChunkSize
+		n, err := inFile.ReadAt(buf, pos)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		// merge this block to what was read so far
+		readData = append(buf[:n], readData...)
+		// count how many lines we have so far, and stop reading if we have
+		// enough
+		foundLines += uint(bytes.Count(buf[:n], []byte{'\n'}))
+		if foundLines >= numLines {
+			break
+		}
+	}
+	// find the start of the n-th to last line
+	slice := readData
+	// `data` contains up to `numLines` lines of the file
+	var data []byte
+	if len(slice) != 0 {
+		if slice[len(slice)-1] == '\n' {
+			// don't consider the last new line for the line count
+			slice = slice[:len(slice)-1]
+		}
+		var (
+			foundLines uint
+			idx        int
+		)
+		for {
+			if foundLines >= numLines {
+				break
+			}
+			// find newlines backwards from the end of `slice`
+			idx = bytes.LastIndexByte(slice, '\n')
+			if idx == -1 {
+				// there are less than `numLines` lines
+				break
+			}
+			foundLines++
+			slice = slice[:idx-1]
+		}
+		if idx == -1 {
+			// if there are less than `numLines` lines, use all what we have read
+			data = readData
+		} else {
+			data = readData[idx+1:] // +1 to skip the newline belonging to the previous line
+		}
+	}
+	// write the requested lines to the writer
+	if _, err = writer.Write(data); err != nil {
+		return err
+	}
+	// reposition the stream at the end, so the caller can keep reading the file
+	// (e.g. when using follow-mode)
+	_, err = inFile.Seek(lastPos, os.SEEK_SET)
+	return err
+}
+
+// readLastLinesFromBeginning reads the last N lines from the provided file,
+// reading from the beginning of the file and keeping track of the last N lines.
+// This is necessary for files that are not seekable (e.g. stdin), but it's less
+// efficient. For an efficient alternative that works on seekable files see
+// readLastLinesBackwards.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile *os.File, writer io.Writer, numLines uint) error {
+	blkSize := getBlockSize(numLines)
+	// read block by block until EOF and store a reference to the last lines
+	buf := make([]byte, blkSize) // will hold each block of read data
+	var (
+		slice      []byte // will hold the final data, after moving line by line
+		foundLines uint
+	)
+	for {
+		n, err := io.ReadFull(inFile, buf)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			if err != io.ErrUnexpectedEOF {
+				return err
+			}
+		}
+		// look for newlines and keep a slice starting at the n-th to last line
+		// (no further than numLines)
+		foundLines += uint(bytes.Count(buf[:n], []byte{'\n'}))
+		slice = append(slice, buf[:n]...) // this is the slice that points to the wanted lines
+		// process the current slice
```

Comment:
I feel like the code here is another candidate for just calling a `lastNLines` function

Reply:
definitely. The code is much shorter now